### ISSUE TITLE
Consider the relevance of custom regularised interpolation gradient

### ISSUE THREAD

**Issue by tvercaut** [DeepRegNet/DeepReg#29]
This is related to #28 and #26 

It seems like autodiff and linear interpolation don't play very nicely together (at least for classical registration approaches). It would be interesting to evaluate how regularised gradients can help and how they would impact learning-based registration as well.

Making even a toy example using autodiff and linear interpolation converge seems indeed complicated:
https://colab.research.google.com/drive/1OmU-jTvWmRRVerc23nsyNz_p8jU3S7wX?usp=sharing

The toy example simply involves registering two slightly offset crops of a 2D image. This is using MSE, translation and BFGS optimiser.

The reason for the convergence issue I think is related to resampling artifacts. deepreg/pytorch/jax currently offers only linear of nearest neighbour resampling, e.g.:
https://pytorch.org/docs/master/nn.functional.html#grid-sample

One option discussed in #26 could be to port the niftyreg-based resampling that had been integrated in niftynet:
https://github.com/NifTK/NiftyNet/tree/dev/niftynet/contrib/niftyreg_image_resampling

Another option is to continue use linear interpolation but implement gradient regularisation. When autograd is used with SSD and linear resampling, you indeed get non negligible local peaks in the cost function and is gradient when the translation parameters are integers.

Obviously, it's not ideal that autograd apparently can't be used directly for classical registration.

In ITK and other classical registration toolkits that do not use autodiff, we always relied on a chain rule for the continuous formulation that eventually implied using the gradient of the moving image (e.g. -2(F-MoT)∇MoT), thereby approximating the gradient of the actual cost function + resampling scheme. This actually has a smoothing/regularisation effect on the gradient which helps avoid local minima at pixel locations.

Note that further regularisation can be obtained by computing ∇MoT using a convolution with the gardient of a Gaussian kernel (see e.g. https://itk.org/Doxygen/html/classitk_1_1DiscreteGaussianDerivativeImageFilter.html)

One could think of implementing a custom interpolation layer that would return such a regularised gradient.

I guess one other option is to change the cost function to use a proper integral rather than a simple resampling the MSE on pixels but that doesn't sound super convenient either. This is briefly discussed in Modersitzki's classical text book (e.g. chapter 6.2):
https://archive.siam.org/books/fa06/Modersitzki_FAIR_2009_FA06.pdf

Adding a bit of noise on the coordinates of the grid points (i.e. some form of dithering) is another form of regularisation wich seems to help.

In terms of further references on this issue. The litterature seems sparse. I saw the paper below earlier but it does not discuss the gradient question:
https://doi.org/10.1109/TIP.2012.2224356

**Comment by mathpluscode** [comment]
@YipengHu Do we pursue this issue?

**Comment by YipengHu** [comment]
closing this now - we will revisit when more evident what to do next.
